sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/ReadFromKafkaDoFn.java

```diff
@@ -376,6 +396,9 @@ public ProcessContinuation processElement(
             outputTimestamp = extractOutputTimestampFn.apply(kafkaRecord);
           }
           receiver.outputWithTimestamp(KV.of(kafkaSourceDescriptor, kafkaRecord), outputTimestamp);
+          if (expectedOffset > endOffset) {
```

Comment:
Since the restriction is already bounded by `endOffset` you don't need to check here as well since `tryClaim` above will return false.